datafusion/optimizer/src/optimizer.rs

```diff
@@ -299,44 +330,71 @@ impl Optimizer {
             log_plan(&format!("Optimizer input (pass {i})"), &new_plan);
 
             for rule in &self.rules {
-                let result =
-                    self.optimize_recursively(rule, &new_plan, config)
-                        .and_then(|plan| {
-                            if let Some(plan) = &plan {
-                                assert_schema_is_the_same(rule.name(), plan, &new_plan)?;
-                            }
-                            Ok(plan)
-                        });
-                match result {
-                    Ok(Some(plan)) => {
-                        new_plan = plan;
-                        observer(&new_plan, rule.as_ref());
-                        log_plan(rule.name(), &new_plan);
-                    }
-                    Ok(None) => {
+                // If skipping failed rules, copy plan before attempting to rewrite
+                // as rewriting is destructive
+                let prev_plan = options
+                    .optimizer
+                    .skip_failed_rules
+                    .then(|| new_plan.clone());
+
+                let starting_schema = new_plan.schema().clone();
+
+                let result = match rule.apply_order() {
+                    // optimizer handles recursion
+                    Some(apply_order) => new_plan.rewrite(&mut Rewriter::new(
```

Comment:
Does rename it to `rewrite_recurisvely` more straightforward?

Reply:
It's a generic api (so using rewrite makes sense to me), it's not introduced by this PR

Reply:
I got a question here.
In previous code, `optimize_inputs` get the `plan.inputs()`. how does `rewrite` get the `plan.inputs()` here? How does `childnode` in `rewrite` equals to `plan.inputs()`

Reply:
Oh, I see the `map_children`

Reply:
Yes, exactly

`rewrite` : https://github.com/apache/arrow-datafusion/blob/cb21404bd3736ff9a6d8d443a67c64ece4c551a9/datafusion/common/src/tree_node.rs#L117-L124

(eventually) calls `map_children`: https://github.com/apache/arrow-datafusion/blob/cb21404bd3736ff9a6d8d443a67c64ece4c551a9/datafusion/common/src/tree_node.rs#L27-L31

This is the beauty of these `TreeNode` APIs and @peter-toth's plan to make them all consistent.